ledger-core/virtual/object/state_report.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2020 Insolar Network Ltd.
+// All rights reserved.
+// This material is licensed under the Insolar License version 1.0,
+// available at https://github.com/insolar/assured-ledger/blob/master/LICENSE.md.
+
+package object
+
+import (
+	"context"
+
+	"github.com/insolar/assured-ledger/ledger-core/conveyor"
+	"github.com/insolar/assured-ledger/ledger-core/conveyor/smachine"
+	"github.com/insolar/assured-ledger/ledger-core/insolar/node"
+	"github.com/insolar/assured-ledger/ledger-core/insolar/payload"
+	"github.com/insolar/assured-ledger/ledger-core/network/messagesender"
+	messageSenderAdapter "github.com/insolar/assured-ledger/ledger-core/network/messagesender/adapter"
+	"github.com/insolar/assured-ledger/ledger-core/pulse"
+	"github.com/insolar/assured-ledger/ledger-core/reference"
+	"github.com/insolar/assured-ledger/ledger-core/vanilla/injector"
+	"github.com/insolar/assured-ledger/ledger-core/vanilla/throw"
+)
+
+type SMReport struct {
+	ReadyToWork smachine.SyncLink
+
+	Reference reference.Global
+
+	report payload.VStateReport
+}
+
+type SMStateReport struct {
+	SMReport
+
+	smachine.StateMachineDeclTemplate
+
+	reportShared SharedReportAccessor
+
+	// dependencies
+	messageSender messageSenderAdapter.MessageSender
+	pulseSlot     *conveyor.PulseSlot
+}
+
+/* -------- Declaration ------------- */
+
+func (sm *SMStateReport) InjectDependencies(stateMachine smachine.StateMachine, _ smachine.SlotLink, injector *injector.DependencyInjector) {
+	s := stateMachine.(*SMStateReport)
+	injector.MustInject(&s.messageSender)
+	injector.MustInject(&s.pulseSlot)
+}
+
+func (sm *SMStateReport) GetInitStateFor(smachine.StateMachine) smachine.InitFunc {
+	return sm.Init
+}
+
+/* -------- Instance ------------- */
+
+func (sm *SMStateReport) GetStateMachineDeclaration() smachine.StateMachineDeclaration {
+	return sm
+}
+
+type SharedReportAccessor struct {
+	smachine.SharedDataLink
+}
+
+func (v SharedReportAccessor) Prepare(fn func(report payload.VStateReport)) smachine.SharedDataAccessor {
+	return v.PrepareAccess(func(data interface{}) bool {
+		fn(*data.(*payload.VStateReport))
+		return false
+	})
+}
+
+type ReportKey struct {
+	ObjectReference reference.Global
+	Pulse           pulse.Number
+}
+
+func BuildReportKey(object reference.Global, pulse pulse.Number) ReportKey {
+	return ReportKey{
+		ObjectReference: object,
+		Pulse:           pulse,
+	}
+}
+
+func GetSharedStateReport(ctx smachine.InOrderStepContext, object reference.Global, pn pulse.Number) (SharedReportAccessor, bool) {
```

Comment:
I suggest to split that into two files:
1. SM
2. Work with shared state

Reply:
done